### PR TITLE
Fix algorithms

### PR DESCRIPTION
**Pull request by vlrsmlkv**
for Mishanya's review ^^

### CODE REVIEW

File: algorithms.js [hunk 1]
```diff
@@ -1,14 +1,14 @@
 // 1. LINEAR SEARCH (returns the index of an array item)
 
-const linearSearch = (array, item) => {
+const linearSearch = (array, key, item) => {
     for (let i = 0; i < array.length; i++) {
-        if (array[i] === item) return i;
+        if ((array[i][key] === item) || (array[i] === item)) {
```
Comment (OUghTMS) [inline algorithms.js]: What about this array?

```
[
  {},
  1,
  {
    pupa: {
      lupa: 1
    }
  },
  0,
  [1, 2, 3]
]
```

And if i gonna search such elements?

```
[1, 2, 3]
```
or
```
{}
```

Will I find something in your opinion?

Comment (OUghTMS) [reply algorithms.js]: how was the key supposed to influence the search?
you do not need to search for something inside the element that you will find in array!
U just need opportunity to find object in array

File: algorithms.js [hunk 2]
```diff
@@ -22,9 +22,9 @@ const binarySearch = (array, item) => {
         if (array[middle] === item) {
             return middle;
         } else if (array[middle] < item) {
-            start = middle +1;
+            start = middle + 1;
```
Comment (OUghTMS) [inline algorithms.js:41]: Good :)

**Review by OUghTMS — CHANGES_REQUESTED** [review]

File: algorithms.js (hunk 2 above)
Comment (vlrsmlkv) [reply algorithms.js:41]: thx

Comment (vlrsmlkv) [reply algorithms.js:41]: that's all i can fix without your further comments...

File: algorithms.js [hunk 3]
```diff
@@ -1,13 +1,43 @@
...
-    for (let i = 0; i < array.length; i++) {
-        if (array[i] === item) return i;
+    if (typeof(item) === "number" || typeof(item) === "string") {
+        for (let i = 0; i < array.length; i++) {
+            if (array[i] === item) return i;
+        }
+
+    } else {
+        for (let i = 0; i < array.length; i++) {
+            if (typeof(array[i]) === "object") {
```
Comment (OUghTMS) [inline algorithms.js]: let's move this into deepEqual

**Pull request merged**